packages/react/src/blocks/ToggleWrapper/ToggleWrapper.tsx

```diff
@@ -0,0 +1,97 @@
+import { Block } from "@blocknote/core";
+import { ReactNode, useEffect, useMemo, useState } from "react";
+
+import { ReactCustomBlockRenderProps } from "../../schema/ReactBlockSpec.js";
+import { useEditorChange } from "../../hooks/useEditorChange.js";
+
+export const ToggleWrapper = (
+  props: Omit<ReactCustomBlockRenderProps<any, any, any>, "contentRef"> & {
+    children: ReactNode;
+    toggledState?: {
+      set: (block: Block<any, any, any>, isToggled: boolean) => void;
+      get: (block: Block<any, any, any>) => boolean;
+    };
+  },
+) => {
+  const { block, editor, children } = props;
+
+  const toggledState = useMemo(
+    () =>
+      props.toggledState || {
+        set: (block, isToggled: boolean) =>
+          window.localStorage.setItem(`toggle-${block.id}`, String(isToggled)),
+        get: (block) =>
+          window.localStorage.getItem(`toggle-${block.id}`) === "true",
+      },
+    [props.toggledState],
+  );
+
+  const [showChildren, setShowChildren] = useState(toggledState.get(block));
+  const [childCount, setChildCount] = useState(block.children.length);
+
+  useEffect(() => {
+    toggledState.set(editor.getBlock(block)!, showChildren);
+  }, [block, block.id, editor, showChildren, toggledState]);
+
+  useEditorChange(() => {
+    const newChildCount = editor.getBlock(block)?.children.length ?? 0;
+
+    if (newChildCount > childCount) {
+      // If a child block is added while children are hidden, show children.
+      if (!showChildren) {
+        setShowChildren(true);
+      }
+    } else if (newChildCount === 0 && newChildCount < childCount) {
+      // If the last child block is removed while children are shown, hide
+      // children.
+      if (showChildren) {
+        setShowChildren(false);
+      }
+    }
+
+    setChildCount(newChildCount);
+  });
```

Comment:
early exit if the thing being wrapped is not even toggleable, otherwise, just overhead.